apps/cdn/src/cache.rs

```diff
@@ -0,0 +1,222 @@
+use std::{
+	hash::Hash,
+	str::FromStr,
+	sync::{atomic::AtomicUsize, Arc},
+};
+
+use reqwest::header::HeaderValue;
+use scc::{ebr::Guard, TreeIndex};
+use scuffle_foundations::http::server::{
+	axum::http::header::{self, HeaderMap, HeaderName},
+	axum::http::{Response, StatusCode},
+	stream::{Body, IntoResponse},
+};
+
+use crate::global::Global;
+
+const ONE_DAY: std::time::Duration = std::time::Duration::from_secs(60 * 60 * 24);
+
+pub struct Cache {
+	inner: moka::future::Cache<CacheKey, CachedResponse>,
+	path_meta: Arc<TreeIndex<String, Arc<PathMeta>>>,
+	http_client: reqwest::Client,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum CacheError {
+	#[error("failed to get cdn file: {0}")]
+	Reqwest(#[from] reqwest::Error),
+	#[error("failed to update path meta index")]
+	PathMetaUpdate,
+}
+
+impl Cache {
+	pub fn new(capacity: u64) -> Self {
+		let path_meta = Arc::new(TreeIndex::new());
+		let pm = Arc::clone(&path_meta);
+
+		Self {
+			inner: moka::future::Cache::builder()
+				.async_eviction_listener(move |key: Arc<CacheKey>, _, _| {
+					let path_meta = Arc::clone(&pm);
+
+					Box::pin(async move {
+						path_meta.remove_async(&key.path).await;
+					})
+				})
+				.expire_after(CacheExpiry)
+				.weigher(|_, v: &CachedResponse| v.body.len() as u32)
+				.max_capacity(capacity)
+				.build(),
+			path_meta,
+			http_client: reqwest::Client::new(),
+		}
+	}
+
+	pub async fn handle_request(
+		&self,
+		global: &Arc<Global>,
+		key: String,
+		mut headers: HeaderMap,
+	) -> Result<CachedResponse, CacheError> {
+		let meta = self.path_meta.peek(&key, &Guard::new()).map(Arc::clone);
+
+		if let Some(meta) = meta {
+			let important_headers = meta
+				.vary_headers
+				.iter()
+				.cloned()
+				.filter_map(|h| headers.remove(&h).map(|v| (h, v)))
+				.collect();
+
+			let cache_key = CacheKey {
+				path: key.clone(),
+				important_headers,
+			};
+
+			if let Some(hit) = self.inner.get(&cache_key).await {
+				hit.hits.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+
+				return Ok(hit);
+			}
+		}
+
+		// miss
+
+		// request file
+		let response = self
+			.http_client
+			.get(format!("{}/{}", global.config.cdn.bucket_url, key))
+			.timeout(std::time::Duration::from_secs(5))
+			.send()
+			.await?;
+
+		if response
+			.headers()
+			.get(header::VARY)
+			.is_some_and(|v| v.to_str().is_ok_and(|v| v == "*"))
+		{
+			// uncacheable
+			return Ok(CachedResponse::from_reqwest(response).await?);
+		}
```

Comment:
Perhaps, store this in the path metadata so that we know its uncacheable. We should also eventually evict it if its there somehow?

Reply:
I'm not sure if that would makes sense. What would be the advantage of storing that? We still have to request origin anyway.

Reply:
If we have request coalescing we would request them sequentially